File: src/main/scala/com/leetcode/SlidingWindowMaximum.scala
package com.leetcode

/**
  * https://leetcode.com/problems/sliding-window-maximum
  */
object SlidingWindowMaximum {

  object Solution {
    import java.util

    def maxSlidingWindow(nums: Array[Int], k: Int): Array[Int] =
      if (nums.isEmpty) Array.empty
      else {
        val windowMax = new Array[Int](nums.length - k + 1)
        val slidingMax = new MaxQueue()

        var i = 0
        while (i < k - 1) {
          slidingMax.enqueue(nums(i))
          i += 1
        }
        while (i < nums.length) {
          slidingMax.enqueue(nums(i))
          windowMax(i - k + 1) = slidingMax.max()
          slidingMax.dequeue()
          i += 1
        }

        windowMax
      }

    private class MaxQueue {
      private val queue = new util.ArrayDeque[Array[Int]]()

      def enqueue(num: Int): Unit = {
        var count = 0
        while (!queue.isEmpty && queue.peekLast()(0) < num) {
          count += queue.pollLast()(1) + 1
        }
        queue.offer(Array(num, count))
      }

      def dequeue(): Unit = {
        val head = queue.peek()
        if (head(1) > 0)
          head(1) -= 1
        else
          queue.poll()
      }

      def max(): Int =
        queue.peek()(0)
    }
  }
}

File: src/main/python/leetcode/sliding_window_maximum.py
# https://leetcode.com/problems/sliding-window-maximum
class Solution(object):
    def maxSlidingWindow(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """

        if not nums:
            return []

        res = []
        deque = MaxQueue()
        k = min(len(nums), k)
        assert k >= 1

        for i in range(0, k - 1):
            deque.enqueue(nums[i])
        for i in range(k - 1, len(nums)):
            deque.enqueue(nums[i])
            res.append(deque.max())
            deque.dequeue()
        return res


class MaxQueue(object):
    def __init__(self):
        from collections import deque
        self.deque = deque()

    def enqueue(self, num):
        cnt = 0
        while len(self.deque) > 0 and self.deque[-1].left < num:
            cnt += self.deque.pop().right + 1
        self.deque.append(Pair(num, cnt))

    def dequeue(self):
        if self.deque[0].right > 0:
            self.deque[0].right -= 1
        else:
            self.deque.popleft()

    def max(self):
        return self.deque[0].left


class Pair(object):
    def __init__(self, left, right):
        self.left = left
        self.right = right
